File: android/ac-rendering/src/main/kotlin/com/microsoft/adaptivecards/rendering/modifiers/ContainerStyleModifier.kt
package com.microsoft.adaptivecards.rendering.modifiers

import androidx.compose.foundation.background
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.microsoft.adaptivecards.core.models.ContainerStyle
import com.microsoft.adaptivecards.hostconfig.LocalHostConfig

/**
 * Applies background color and corner radius based on container style and host config.
 */
@Composable
fun Modifier.containerStyle(
    style: ContainerStyle?,
    cornerRadiusDp: Int? = null
): Modifier {
    if (style == null && cornerRadiusDp == null) {
        return this
    }

    val hostConfig = LocalHostConfig.current
    var result = this

    // Apply corner radius
    val radius = cornerRadiusDp ?: hostConfig.cornerRadius.container
    if (radius > 0) {
        result = result.clip(RoundedCornerShape(radius.dp))
    }

    // Apply background color from container style
    if (style != null) {
        val styleConfig = when (style) {
            ContainerStyle.Default -> hostConfig.containerStyles.default
            ContainerStyle.Emphasis -> hostConfig.containerStyles.emphasis
            ContainerStyle.Good -> hostConfig.containerStyles.good
            ContainerStyle.Attention -> hostConfig.containerStyles.attention
            ContainerStyle.Warning -> hostConfig.containerStyles.warning
            ContainerStyle.Accent -> hostConfig.containerStyles.accent
        }
        val backgroundColor = parseColor(styleConfig.backgroundColor)
        result = result.background(backgroundColor)
    }

    return result
}

/**
 * Parse color string to Compose Color, supporting #AARRGGBB and #RRGGBB formats.
 */
internal fun parseColor(colorString: String): Color {
    return try {
        Color(android.graphics.Color.parseColor(colorString))
    } catch (e: Exception) {
        Color.White
    }
}

File: ios/Sources/ACRendering/Modifiers/ContainerStyleModifier.swift
import SwiftUI
import ACCore

public extension View {
    /// Applies container style background color and corner radius
    func containerStyle(_ style: ContainerStyle?, hostConfig: HostConfig) -> some View {
        self.modifier(ContainerStyleModifier(style: style, hostConfig: hostConfig))
    }
}

struct ContainerStyleModifier: ViewModifier {
    let style: ContainerStyle?
    let hostConfig: HostConfig

    func body(content: Content) -> some View {
        content
            .background(
                RoundedRectangle(cornerRadius: cornerRadius)
                    .fill(backgroundColor)
            )
            .clipShape(RoundedRectangle(cornerRadius: cornerRadius))
    }

    private var backgroundColor: Color {
        guard let containerStyle = style else { return .clear }
        let styleConfig: ContainerStyleConfig

        switch containerStyle {
        case .default:
            styleConfig = hostConfig.containerStyles.default
        case .emphasis:
            styleConfig = hostConfig.containerStyles.emphasis
        case .good:
            styleConfig = hostConfig.containerStyles.good
        case .attention:
            styleConfig = hostConfig.containerStyles.attention
        case .warning:
            styleConfig = hostConfig.containerStyles.warning
        case .accent:
            styleConfig = hostConfig.containerStyles.accent
        }

        return Color(hex: styleConfig.backgroundColor)
    }

    private var cornerRadius: CGFloat {
        guard let containerStyle = style else { return 0 }
        let key: String

        switch containerStyle {
        case .default:
            key = "container"
        case .emphasis:
            key = "container"
        case .good:
            key = "container"
        case .attention:
            key = "container"
        case .warning:
            key = "container"
        case .accent:
            key = "container"
        }

        return CGFloat(hostConfig.cornerRadius[key] ?? 0)
    }
}
